DevWs for Plaid Systems LLC containing containing Spruce Controller SST

natec007 opened the pull request.

natec007 committed:
DevWs for Plaid Systems LLC containing containing Spruce Controller SST

natec007 committed:
Add Spruce Valve which is the required child device for Spruce Controller

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,428 +10,368 @@
+        command "config"
+        command "refresh"        
+        command "rain"
+        command "settingsMap"
```
Did you defined these commands in your custom capabilities?

natec007 replied:
Ok, removed unused commands.  Commands listed are custom capability or the standard "switch" commands

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,428 +10,368 @@
+
 		//new release
-        fingerprint endpointId: "1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18", profileId: "0104", deviceId: "0002", deviceVersion: "00", inClusters: "0000,0003,0004,0005,0006,0009,000A,000F", outClusters: "0003, 0019", manufacturer: "PLAID SYSTEMS", model: "PS-SPRZ16-01", deviceJoinName: "Spruce Irrigation"
+        fingerprint endpointId: "1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18", profileId: "0104", deviceId: "0002", deviceVersion: "00", inClusters: "0000,0003,0004,0005,0006,000F", outClusters: "0003, 0019", manufacturer: "PLAID SYSTEMS", model: "PS-SPRZ16-01", deviceJoinName: "Spruce Controller"
```
Please leave old deviceJoinName. We've decided to keep devices' names across whole system more unified, so they should follow `<manufacturer> + <device type>`

natec007 replied:
deviceJoinName set to "Spruce Irrigation Controller (Gen1)"

PKacprowiczS reviewed (changes requested):
Can you update your code, so it would follow a specified coding style? Something like described here: https://github.com/greens/DTH_development

natec007 committed:
Update code style, remove undefined commands, update fingerprint syntax, deprecate SmartApp scheduler compatibility and replace with settings that allow standard automation use

natec007 commented:
Pushed to the devws-for-plaid-systems-llc-cont_6 branch on natec007/SmartThingsPublic.  Is there another step to request pull through the developer portal?
Removed Valve Slider from presentation since it has stopped working as reported through forum and confirmed by Nayelyz https://community.smartthings.com/t/custom-capability-and-cli-developer-preview/197296/851 -will add back into presentation when resolved.
thanks

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,476 +10,491 @@
-        
-		//ST release
-		//fingerprint endpointId: '1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18', profileId: '0104', deviceId: '0002', deviceVersion: '00', inClusters: '0000,0003,0004,0005,0006,000F', outClusters: '0003, 0019', manufacturer: 'PLAID SYSTEMS', model: 'PS-SPRZ16-01'
+	definition (name: "Spruce Controller", namespace: "plaidsystems", author: "Plaid Systems", mnmn: "SmartThingsCommunity", vid: "566e4d4f-93bc-3c86-b7e1-8a4e212683b0"){//vid: "5e1426e6-1943-3331-8880-6c3e7743f2cc"){
```
Can you give me insights how did you defined this vid? Did you used SmartThings CLI for this?

natec007 replied:
Yes, SmartThings CLI.  Newest commit has an updated vid with new presentation, because previous presentation had a duplicate rainSensor component I had put in to try and force a presentation update previously.

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,476 +10,491 @@
 		//new release
-        fingerprint endpointId: "1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18", profileId: "0104", deviceId: "0002", deviceVersion: "00", inClusters: "0000,0003,0004,0005,0006,0009,000A,000F", outClusters: "0003, 0019", manufacturer: "PLAID SYSTEMS", model: "PS-SPRZ16-01", deviceJoinName: "Spruce Irrigation"
-		
+        fingerprint manufacturer: "PLAID SYSTEMS", model: "PS-SPRZ16-01", deviceJoinName: "Spruce Irrigation Controller (Gen1)"
```
please remove (Gen1)

natec007 replied:
Ok, removed Gen1

KKlimczukS replied:
@natec007 what about this one ? "(Gen1)" is still there.

PKacprowiczS reviewed (changes requested):
Hey @natec007 
While you have chance to work on this code, I'd make it more read friendly - once again I'd like to ask you to follow this guide https://github.com/greens/DTH_development and correct spacing and indentation in the code.
Also I'd recommend getting rid of any magic strings and numbers and replace them with correctly named constants. Where possible, I'd also use zigbee commands methods (like `zigbee.writeAttribute(cluster, attribute, value)` instead of commands written as strings

natec007 committed:
updated zigbee commands, constants, whitespace, tab spacing

natec007 committed:
Updated vid to remove a (temporary) duplicate entry in the json

natec007 commented:
Updated with constants and zigbee commands so values use constant and datatypes.  Removed whitespace, checked it is tabs indentation.  Also the slider function was fixed, so presentation for Valve has been updated to use the slider for changing the duration in minutes.
thanks

PKacprowiczS commented on devicetypes/plaidsystems/spruce-valve.src/spruce-valve.groovy:
```diff
@@ -0,0 +1,76 @@
+	sendEvent(name: "DeviceWatch-Enroll", value: [protocol: "cloud", scheme:"untracked"].encodeAsJson(), displayed: false)
+}
+
+def parse(String onOff) {
```
Is it ever called?

PKacprowiczS replied:
I need to ask again, what's the idea behind this? No command is coming directly to child device, so I don't know why `parse()` is declared here. Can you elaborate?
Is it just because you're using `child.sendEvent(name: "valve", value: onoff)` in your main DTH? If so, you don't need any "passthrough" in child DTH, this event will be registered for it right away.

PKacprowiczS commented on devicetypes/plaidsystems/spruce-valve.src/spruce-valve.groovy:
```diff
@@ -0,0 +1,76 @@
+}
+
+def open() {
+	def eventMap = createEvent(dni: device.deviceNetworkId, value: 'open', duration: device.latestValue("valveDuration").toInteger(), label: device.label)
```
Is it necessary to create an event here? Just a map wouldn't sufficed?

KKlimczukS replied:
@natec007 Please, could You answer Przemek's question ?

PKacprowiczS reviewed (commented):
Thank you, that looks way better. If it works well, I don't see any major issue, I could still nitpick about some spacing.
@tpmanley @greens could you have a look?

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
+
+//parse hex string and make integer
+private hextoint(String hex) {
+	Long.parseLong(hex, 16).toInteger()
```
Any particular values you're expecting here that you didn't want to just use `Integer.parseInt()` so you could avoid the cast?

PKacprowiczS replied:
Bumping Steven's comment.

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
+def updated() {
+	log.debug "updated"
+
+	createChildDevices()
```
can you wrap this in an

`if (!childDevices)` just to avoid possible messy behavior from overwrite attempts?

PKacprowiczS replied:
He avoids it here https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/47339/files#diff-203530982adf372057d9eaac68564e740dc55e3cc174c90d1d60ad524ec56834R227

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
+						[completedSetup: true, label: "${childLabel}",
+						 isComponent: false])
+						 log.debug "${child}"
+					child.sendEvent(name: "switch", value: "off", displayed: false)
```
nit: whitespace is confusing here.

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
+
+	def endpointMap = [:]
+	int zone = 0
+	while(zone <= 17)
```
did you want to make this a `for...in` loop while you're here?

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
+def getVERSION() {'v3.2 11-2020'}
+def getDEBUG() {false}
+//zigbee cluster, attribute, identifiers
+def getON_OFF_CLUSTER() {0x0006}
```
this one at least is already defined for you in `zigbee.ONOFF_CLUSTER`

PKacprowiczS replied:
I'm with Steven on this. Here's a list of constants defined in `zigbee`

```
zigbee.BASIC_CLUSTER = 0x0000
zigbee.POWER_CONFIGURATION_CLUSTER = 0x0001
zigbee.ONOFF_CLUSTER = 0x0006
zigbee.LEVEL_CONTROL_CLUSTER = 0x0008
zigbee.COLOR_CONTROL_CLUSTER = 0x0300
zigbee.IAS_ZONE_CLUSTER = 0x0500
zigbee.TEMPERATURE_MEASUREMENT_CLUSTER = 0x0402
zigbee.RELATIVE_HUMIDITY_CLUSTER = 0x0405
zigbee.SIMPLE_METERING_CLUSTER = 0x0702
zigbee.ELECTRICAL_MEASUREMENT_CLUSTER = 0x0B04


zigbee.ATTRIBUTE_ACTIVE_POWER = 0x050B
zigbee.ATTRIBUTE_INSTANTANEOUS_POWER_DEMAND = 0x0400
zigbee.ATTRIBUTE_COLOR_TEMPERATURE = 0x0007
zigbee.ATTRIBUTE_IAS_ZONE_STATUS = 0x0002
zigbee.ATTRIBUTE_IAS_ZONE_CIE_ADDRESS = 0x0010
```

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
+	sendEvent(name: "healthStatus", value: "online")
+	sendEvent(name: "DeviceWatch-Enroll", value: 2* 60 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+
+	configureHealthCheck()
```
This is no longer necessary. 

just sending

`[name: "checkInterval", value: hcIntervalMinutes * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID]]`

and having a defined `ping` method should get you what you want. You can also send a similar event for all the child devices as long as they each have a ping method defined as well and then everything will be checked automatically if we haven't heard anything from the device itself.

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
+	//setup reporting for 18 endpoints
+	def reportingCmds = []
+	reportingCmds += zigbee.configureReporting(ON_OFF_CLUSTER, 0, DataType.BOOLEAN, 1, 0, 0x01, [destEndpoint: 1])
+	reportingCmds += zigbee.configureReporting(ALARMS_CLUSTER, 0, DataType.UINT16, 1, 0, 0x00, [destEndpoint: 1])
```
`zigbee.configureReporting` includes the binding command as well so the `addBinding` commands above are redundant.

KKlimczukS replied:
@natec007 Please, could You change the code according to Steven's suggestions ?

greens commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,570 @@
-def z16off() {
-	"st cmd 0x${device.deviceNetworkId} 17 6 0 {}"    
+
+private String swapEndianHex(String hex) {
```
You should be able to achieve what you want here using the builtins:

`zigbee.swapEndianHex(zigbee.convertToHexString(...))`

greens commented on devicetypes/plaidsystems/spruce-valve.src/spruce-valve.groovy:
```diff
@@ -0,0 +1,76 @@
+		capability "heartreturn55003.valveDuration"
+
+		command "open"
+		command "close"
```
`open` and `close` are part of the `Valve` capability, and so these are redundant.

PKacprowiczS replied:
Just bumping up Steven's comment.

KKlimczukS replied:
@natec007 Please change the code according to Steven's comment.

greens commented on devicetypes/plaidsystems/spruce-valve.src/spruce-valve.groovy:
```diff
@@ -0,0 +1,76 @@
+
+	sendEvent(name: "DeviceWatch-DeviceStatus", value: "online")
+	sendEvent(name: "healthStatus", value: "online")
+	sendEvent(name: "DeviceWatch-Enroll", value: [protocol: "cloud", scheme:"untracked"].encodeAsJson(), displayed: false)
```
see comments about this on parent DTH

PKacprowiczS replied:
Bump. Defining `checkInterval` should be enough, since parent's `ping()` is taking care of polling the device.

natec007 replied:
Does child need any device health since parent takes care of this?  Can I remove capability "Health Check" and associated code.  And if I need to keep it, do I use checkInterval or the DeviceWatch-Enroll which sets it to cloud/untracked?

KKlimczukS replied:
I don't think You need DeviceWatch-Enroll / untracked here. 
Try to leave "Health Check" here and redirect ping() calls to partent DTH -> 

def ping() {
	refresh()
}

def refresh() {
	// You will have to implement this method in spruce-controller
	parent.refreshChild()
       // or just call parent.refresh() if it already polls all the endpoints
 }

KKlimczukS replied:
additionally: https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/47339#discussion_r518332971

greens reviewed (commented):
@natec007 I'm a bit concerned that the migration between these versions might not go smoothly.

Also, as you're most likely aware, the way this is written now will create 17 different device entries in the main device list, leaving it a bit cluttered. While we don't yet have full support for addressing multi-component devices in the metadata generation tool, might I instead suggest that the old approach of using custom commands and attributes for each zone might allow you to have a better presentation? You could add them to your custom capability and generate metadata for their presentation.

natec007 commented:
thanks @PKacprowiczS 
@greens is there a guideline on how to get the presentation to be arranged in a specific order?  I had tested out some presentations similar to your recommendation, and the order of the components was not correct.  They would present starting with zone10-16, then go back to zone1-9, and I believe the status was also in the wrong location or at the bottom of the list for some reason.

The second issue was that automations are not supported for custom capabilities, severely limiting usefulness, was this going to be implemented in the future?
thanks

greens commented:
@natec007 let me see if I can get some more answers for you.

dkirker commented:
@natec007 The items *should* show up in the order that they are in the metadata. I'll see if I can pull down the metadata this afternoon and have a look at it (you're also welcome to email it to me if you have a local copy). I've been working to familiarize myself with the new APIs for metadata and how those work with custom capabilities and components. I'll see if I can help you out.

One thing to note re attributes, right now the plugin doesn't show more than one attribute per capability... This affects standard capabilities as well as custom, though most standard capabilities have logic internal in the plugin for how to render it. There is a request to get this changed.

natec007 commented:
@dkirker @greens thanks, I can try it again and then send it to you.  I may be mistaken, I had tried using children as components similar to the remote control devices and those definitely were in a different order, but maybe the metadata presentation are ok, will try it again.

KKlimczukS commented:
@natec007 Please, could You share any update on this pull request ?

natec007 commented:
@KKlimczukS I am still waiting on an update as to how the 16 different zones can be propagated into a single presentation, and was also wondering when they would be available for use in automations.  @dkirker had provided some info but it sounded like there are still some things in development that may be needed.  Do I need to create 16 different capabilities to present and control them, or can they be put into a single capability?  I was not able to get the presentation to work when using them in 1 capability. Also using separate components does not seem to present or I could be missing something.  Welcome to suggestions.

dkirker commented:
@natec007 I think we'll go the multi-component route. I will follow up to your email with steps that we need to take. Using different capabilities doesn't really make much sense, and is currently held up with updates being needed to the plugin.

natec007 commented:
Thanks @dkirker !  Sounds good, I think I'm still not sure how to go about the multi-component and have them display properly.  So the extra steps will be handy to move it along.  Will check up through email.

PKacprowiczS commented:
@natec007 do you need any additional help? I also might be able to give you some insights.
Maybe you'd add me to yours and @dkirker's e-mail loop? Of course if you both don' mind.

natec007 committed:
Single presentation Spruce Controller update

natec007 commented:
@PKacprowiczS I just pushed single presentation.  Yes any insights would be great.  What email address?  Thanks

PKacprowiczS commented:
@natec007 I DMed you my e-mail adress on ST community.

natec007 commented:
@PKacprowiczS Have you had a chance to review the updated single instance version?
thanks

natec007 committed:
Update Spruce Controller with patch fix for automations and some other fixes

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
+    	ocfDeviceType: "x.com.st.d.remotecontroller", mcdSync: true, vid: "2914a12b-504f-344f-b910-54008ba9408f") {
+
+        capability "Actuator"
+		capability "Switch"
```
Those indentations are on purpose? I think it'll look better when properly adjusted.

natec007 replied:
ok, should now be fixed

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
+	def map = zigbee.parseDescriptionAsMap(description)
+	if (DEBUG && !map.raw) log.debug "map ${map}"
+
+	if (description.contains("on/off")){
```
nit: lacking space between `){` (please change every occurence)

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
-			return result
+
+	if (DEBUG && command != null) log.debug "${command} >> endpoint ${endpoint} value ${value} cluster ${map.clusterInt}"
+	switch(command) {
```
nit: lacking space between `switch` and `(command)

natec007 replied:
ok, should now be fixed

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
+		if (scheduleState == "pause") log.debug "pausing schedule"
+		else {
+			if (scheduleStatus != "off" && scheduleValue == "off") result.push(createEvent(name: "status", value: "Schedule ${scheduleValue}"))
+			result.push(createEvent(name: "controllerState", value: scheduleValue))
```
Including only this one to the `if` statement above is on purpose? If so, please adjust line below.

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
+
+	//get and delete children avoids duplicate children
+	def children = getChildDevices()
+	if(children != null){
```
nit: space after `if`

natec007 replied:
ok, should now be fixed

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
+	if (DEBUG) log.debug "Pump/Master: ${pumpMasterEndpoint} splitCycle: ${splitCycle} valveDelay: ${valveDelay}"
+
+	def endpointMap = [:]
+    for (zone in 0..17) {
```
nit: spaces instead of tab

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
+	for (i in 1..16){
+		def endpoint = i + 1
+		//if (settings."z${i}" && settings."z${i}Duration" != null){
+        if (settings."z${i}Duration" != null){
```
nit: spaces instead of tabs

PKacprowiczS commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,553 @@
-def z16on() {
-	return manual() + "st cmd 0x${device.deviceNetworkId} 17 6 1 {}"    
+
+def healthPoll() {
```
Instead of this method, you should have `ping()` method implemented.

Generally speaking, you don't need any code related to keeping device online. 
1) Include `capability "Health Check"`, 
2) Define checkInterval with
`[name: "checkInterval", value: hcIntervalMinutes * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID]]`
3) Implement `ping()` method.

natec007 replied:
ok, changed to ping and checkInterval is in configure

PKacprowiczS reviewed (changes requested):
I think you've missed a few insights from Steven, please have a look at them.

natec007 committed:
update format, and health check

PKacprowiczS commented on devicetypes/plaidsystems/spruce-valve.src/spruce-valve.groovy:
```diff
@@ -0,0 +1,67 @@
+**/
+
+metadata {
+	definition (name: "Spruce Valve", namespace: "plaidsystems", author: "Plaid Systems", mnmn: "SmartThingsCommunity") {//, vid: "558e73d3-c800-3669-b276-1d4352eda12b"){
```
If you're not using this vid here, remove it completely.

natec007 committed:
Update hextoint, remove health check from child

natec007 commented:
Removed health check from child device, fixed hextoint to use built in function, removed vid and unnecessary code.

PKacprowiczS reviewed (dismissed):
@greens @dkirker @KKlimczukS do you mind taking another look?

dkirker commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,525 @@
-    preferences {
-    input description: 'If you have a rain sensor wired to the rain sensor input on the Spruce controller, turn it on here.', displayDuringSetup: true, type: 'paragraph', element: 'paragraph', title: 'Rain Sensor'
-    input description: 'The SYNC SETTINGS button must be pressed after making a change to the Rain sensor:', displayDuringSetup: false, type: 'paragraph', element: 'paragraph', title: ''     
-    input 'RainEnable', 'bool', title: 'Rain Sensor Attached?', required: false, displayDuringSetup: true
```
Since this variable name was changed, is there any logic that will also honor the old name so that things don't break for old installations?

natec007 replied:
I understand the concern, but I'm not sure this is an issue, since the old installations are already broken after the release of the new app.  The device is not useable in the new app since none of the valves are displayed and the smartapp is also no longer available in ST.

This updated dth no longer uses the smartapp and is automations friendly since it uses the child devices.  I think this makes it a better fit overall for the ST platform.

dkirker commented on devicetypes/plaidsystems/spruce-controller.src/spruce-controller.groovy:
```diff
@@ -14,736 +10,525 @@
+	bindCmds += zigbee.addBinding(ALARMS_CLUSTER, [destEndpoint: 1])
+
+	for (endpoint in 1..18) {
+		bindCmds += zigbee.addBinding(BINARY_INPUT_CLUSTER, [destEndpoint: endpoint])
```
Just a note re the comment below: https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/47339/files?diff=unified&w=1#r518333854

The `zigbee.configureReporting` internally calls `addBinding(cluster, additionalParams)`, so these packets will be sent twice. `configure()` isn't something that happens often, but there is always the chance that the extra packets could tie things up or get dropped (especially if the link to the device is weak), or cause issues for the Zigbee network if there are many devices on it.

natec007 replied:
Ok, removed bindCmds and associated code, since reporting handles this

dkirker reviewed (dismissed):
I think that is probably all good. Added a couple of notes. One is about continuity for existing integrations, and the other is just re mitigation of a packet storm.

natec007 committed:
remove binding since included in reporting configuration

natec007 commented:
@dkirker removed the extra binding cmds, @KKlimczukS I think that was old rev, not sure where that came from as the updated valve has the corrections requested.
